### PR TITLE
Feature/addons

### PR DESCRIPTION
**Pull request by Npwskp**
review duay

### CODE REVIEW

File: src/schema/addons.schema.ts [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
   schemaOptions: { collection: 'addons' },
 })
 export class AddonSchema {
   _id: Types.ObjectId;
+
+  @Prop({ required: true })
+  @ApiProperty()
+  title: string;
+
+  @Prop()
```
Comment (ishiasaka) [inline src/schema/addons.schema.ts]: Add `{required: true}`

File: src/addons/addons.controller.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+@Controller('addons')
+@ApiTags('addons')
+export class AddonsController {
+  constructor(private readonly addonService: AddonsService) {}
+
+  @ApiResponse({
+    status: HttpStatus.CREATED,
+    description: 'Create addon',
+  })
+  @Post()
```
Comment (ishiasaka) [inline src/addons/addons.controller.ts:29]: Post should return 201 (Created)

**Review by ishiasaka — CHANGES_REQUESTED** [review]

File: src/addons/addons.controller.ts (hunk 2 above)
Comment (Npwskp) [reply src/addons/addons.controller.ts:29]: done

File: src/schema/addons.schema.ts (hunk 1 above)
Comment (Npwskp) [reply src/schema/addons.schema.ts]: done

File: src/schema/addons.schema.ts [hunk 3]
```diff
@@ -1,9 +1,22 @@
-import { modelOptions } from '@typegoose/typegoose';
+import { ApiProperty } from '@nestjs/swagger';
+import { Prop, modelOptions } from '@typegoose/typegoose';
 import { Types } from 'mongoose';
 
 @modelOptions({
   schemaOptions: { collection: 'addons' },
 })
 export class AddonSchema {
   _id: Types.ObjectId;
```
Comment (lazykern) [inline src/schema/addons.schema.ts:11]: Sai swagger decorators hai field nee duay krub kuay

Tua yang khong au:
```typescript
  @Prop({ auto: true })
  @ApiProperty({ type: String, description: 'Session ID' })
  _id: Types.ObjectId;
```

File: src/schema/addons.schema.ts [hunk 4]
```diff
@@ -1,9 +1,22 @@
...
+  @ApiProperty()
+  title: string;
+
+  @Prop({ required: true })
+  @ApiProperty()
+  price: number;
+
+  @Prop({ default: null })
+  @ApiProperty()
+  delete_at: Date;
```
Comment (lazykern) [inline src/schema/addons.schema.ts]: Must be `deleted_at`

**Review by lazykern — CHANGES_REQUESTED** [review]

File: src/schema/addons.schema.ts (hunk 3 above)
Comment (Npwskp) [reply src/schema/addons.schema.ts:11]: do i need to do this with category too??

Comment (Npwskp) [reply src/schema/addons.schema.ts:11]: btw what does the description: 'Session ID' do

Comment (lazykern) [reply src/schema/addons.schema.ts:11]: > do i need to do this with category too??

Yes, for every api property. Cuz swagger won't document it if there's no decorator on the field

Comment (lazykern) [reply src/schema/addons.schema.ts:11]: > btw what does the description: 'Session ID' do

For a sexy schema documentation (well it's what @hirasawaau wrote)
 
image

Comment (lazykern) [reply src/schema/addons.schema.ts:11]: FYI

image

File: src/addons/addons.controller.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  createAddon(@Body() doc: CreateAddonDto) {
+    return this.addonService.createAddon(doc.title, doc.price);
+  }
+
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description: 'Get all addons',
```
Comment (ishiasaka) [inline src/addons/addons.controller.ts:36]: Please specific type of response

File: src/addons/addons.controller.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  })
+  @Get()
+  getAllAddons() {
+    return this.addonService.getAllAddons();
+  }
+
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description: 'Get addon by Id',
```
Comment (ishiasaka) [inline src/addons/addons.controller.ts:47]: Please specific type of response

File: src/addons/addons.controller.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  })
+  @Get(':id')
+  getAddon(@Param('id') id: string) {
+    return this.addonService.getAddonById(id);
+  }
+
+  @ApiResponse({
+    status: HttpStatus.OK,
+    description: 'Update addon',
```
Comment (ishiasaka) [inline src/addons/addons.controller.ts]: Please specific type of response

File: src/schema/addons.schema.ts [hunk 8]
```diff
@@ -1,9 +1,24 @@
...
+  @Prop({ required: true })
+  @ApiProperty()
+  title: string;
+
+  @Prop({ required: true })
+  @ApiProperty()
+  price: number;
+
+  @Prop({ default: null })
+  @ApiProperty()
```
Comment (ishiasaka) [inline src/schema/addons.schema.ts]: Don't export `delete_at` to response

**Review by ishiasaka — CHANGES_REQUESTED** [review]

File: src/addons/dto/addon.dto.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+export class CreateAddonDto {
+  @ApiProperty()
+  title: string;
```
Comment (ishiasaka) [inline src/addons/dto/addon.dto.ts:7]: Please use `class-validator` `@IsString()`

File: src/addons/dto/addon.dto.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+
+  @ApiProperty()
+  price: number;
```
Comment (ishiasaka) [inline src/addons/dto/addon.dto.ts:11]: `@IsNumber()`

**Review by ishiasaka — CHANGES_REQUESTED** [review]

File: src/addons/addons.service.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    return await this.addonModel
+      .findOneAndUpdate({ _id: id, deleted_at: null }, updateAddon, {
+        new: true,
+      })
+      .exec();
+  }
+
+  deleteAddon(id: string) {
+    return this.addonModel
+      .updateOne({ _id: id }, { deleted_at: new Date() })
```
Comment (ishiasaka) [inline src/addons/addons.service.ts:39]: Add `.orFail()`

**Review by ishiasaka — CHANGES_REQUESTED** [review]

**Review by ishiasaka — APPROVED** [review]

**Review by lazykern — APPROVED** [review]

**Pull request merged**